crates/ra_assists/src/handlers/change_visibility.rs

```diff
@@ -72,6 +77,125 @@ fn add_vis(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+            target_name = Some(adt.name(ctx.db));
+            match adt {
+                hir::Adt::Struct(s) => offset_target_and_file_id(ctx.db, s),
+                hir::Adt::Union(u) => offset_target_and_file_id(ctx.db, u),
+                hir::Adt::Enum(e) => offset_target_and_file_id(ctx.db, e),
+            }
+        }
+        hir::ModuleDef::Const(c) => {
+            target_name = c.name(ctx.db);
+            offset_target_and_file_id(ctx.db, c)
+        }
+        hir::ModuleDef::Static(s) => {
+            target_name = s.name(ctx.db);
+            offset_target_and_file_id(ctx.db, s)
+        }
+        hir::ModuleDef::Trait(t) => {
+            target_name = Some(t.name(ctx.db));
+            offset_target_and_file_id(ctx.db, t)
+        }
+        hir::ModuleDef::TypeAlias(t) => {
+            target_name = Some(t.name(ctx.db));
+            offset_target_and_file_id(ctx.db, t)
+        }
+        hir::ModuleDef::Module(m) => {
+            target_name = m.name(ctx.db);
+            let in_file_source = m.declaration_source(ctx.db)?;
+            let file_id = in_file_source.file_id.original_file(ctx.db);
+            let syntax = in_file_source.value.syntax();
+            (vis_offset(syntax), syntax.text_range(), file_id)
+        }
+        // Enum variants can't be private, we can't modify builtin types
+        hir::ModuleDef::EnumVariant(_) | hir::ModuleDef::BuiltinType(_) => return None,
+    };
+
+    // FIXME if target is in another crate, add `pub` instead of `pub(crate)`
+
+    let assist_label = match target_name {
+        None => "Change visibility to pub(crate)".to_string(),
+        Some(name) => format!("Change visibility of {} to pub(crate)", name),
+    };
+
+    acc.add(AssistId("change_visibility"), assist_label, target, |edit| {
+        edit.set_file(target_file);
+        edit.insert(offset, "pub(crate) ");
+        edit.set_cursor(offset);
+    })
+}
+
+fn add_vis_to_referenced_record_field(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let record_field: ast::RecordField = ctx.find_node_at_offset()?;
+    let (record_field_def, _) = ctx.sema.resolve_record_field(&record_field)?;
+
+    let current_module = ctx.sema.scope(record_field.syntax()).module()?;
+    let visibility = record_field_def.visibility(ctx.db);
+    if visibility.is_visible_from(ctx.db, current_module.into()) {
+        return None;
+    }
+
+    let in_file_source = record_field_def.source(ctx.db);
+    let (offset, target) = match in_file_source.value {
+        hir::FieldSource::Named(it) => {
+            let s = it.syntax();
+            (vis_offset(s), s.text_range())
+        }
+        hir::FieldSource::Pos(it) => {
+            let s = it.syntax();
+            (vis_offset(s), s.text_range())
+        }
+    };
+
+    let target_file = in_file_source.file_id.original_file(ctx.db);
+
+    acc.add(
+        AssistId("change_visibility"),
+        "Change visibility of {}.{} to pub(crate)",
```

Comment:
I missed something here, woops.

Reply:
Hmm... adding the same crate check requires finding out the target module, which i didn't need to do so far. I don't see a way to get the module of a `hir::Field`, so I tried `ctx.sema.scope(&the_field_syntax).module()?`, and that fails with FieldDefs in other files :/
Any tips?

Reply:
I think `record_field_def.parent_def().module()` should be it :slightly_smiling_face:

Reply:
Well... that was visible :sweat_smile: thanks!